### PR TITLE
c8d: Handle userns properly

### PR DESCRIPTION
**Pull request by rumpl**
**- What I did**

Remapped the GID/UID when a snapshot is created if the daemon is started with `--userns-remap`. This implementation differs from the graph driver one slightly. With graph drivers the files/directories are chown'ed when the image is downloaded and the contents extracted. Basically, with graph drivers we remap on pull and with containerd we remap on run.

Note: with this change builds still fail, my guess is because buildkit still mounts the roots in normal mode, I'll continue looking but I think we can merge this one, I am pretty sure some changes in buildkit are needed:

```console
$ docker build -t userns . 
[+] Building 1.7s (6/6) FINISHED                                                                                                                               docker:local
 => [internal] load build definition from dd                                                                                                                           0.0s
 => => transferring dockerfile: 71B                                                                                                                                    0.0s
 => [internal] load .dockerignore                                                                                                                                      0.0s
 => => transferring context: 2B                                                                                                                                        0.0s
 => [internal] load metadata for docker.io/library/debian:bullseye-slim                                                                                                1.3s
 => [auth] library/debian:pull token for registry-1.docker.io                                                                                                          0.0s
 => CACHED [1/2] FROM docker.io/library/debian:bullseye-slim@sha256:61386e11b5256efa33823cbfafd668dd651dbce810b24a8fb7b2e32fa7f65a85                                   0.0s
 => => resolve docker.io/library/debian:bullseye-slim@sha256:61386e11b5256efa33823cbfafd668dd651dbce810b24a8fb7b2e32fa7f65a85                                          0.0s
 => ERROR [2/2] RUN touch testa                                                                                                                                        0.3s
------
 > [2/2] RUN touch testa:
0.287 runc run failed: unable to start container process: error during container init: error mounting "/tmp/root/165536.165536/buildkit/executor/hosts.iv21cdui6qvdwqel9cdo93spo" to rootfs at "/etc/hosts": open /tmp/root/165536.165536/buildkit/executor/7yecjv2i25qdu44khgx41b450/rootfs/etc/hosts: permission denied
------
dd:2
--------------------
   1 |     FROM debian:bullseye-slim
   2 | >>> RUN touch testa
   3 |     
--------------------
ERROR: failed to solve: process "/bin/sh -c touch testa" did not complete successfully: exit code: 1
```

**- How I did it**

**- How to verify it**

`TestDaemonUserNamespaceRootSetting` should not fail any more
 
```console
$ make DOCKER_GRAPHDRIVER=overlayfs TEST_FILTER=TestDaemonUserNamespaceRootSetting TEST_INTEGRATION_USE_SNAPSHOTTER=1 test-integration
```

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Review by vvoland — COMMENTED** [review]
Does this break `diff`/`commit` (all files are chowned and treated as modified)?

EDIT: Ah, it shouldn't because it creates 2 snapshots on top of the base image.
Wondering if we should also adjust `calculateSnapshotParentUsage` to account for that?

**Comment by rumpl** [comment]
> Wondering if we should also adjust `calculateSnapshotParentUsage` to account for that?

Hm... With overlayfs this would mean the reported size is 2x if metacopy is not enabled, which is true since chown would copy the files.

**Comment by thaJeztah** [comment]
Looking forward to idmapped mounts, which probably would make these kind of things a lot easier.

File: daemon/containerd/image_snapshot_unix.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	}); err != nil {
+		return err
+	}
+
+	mounts, err := snapshotter.Prepare(ctx, remappedID, parentSnapshot)
+	if err != nil {
+		return err
+	}
+
+	if err := remapRootFS(ctx, mounts, rootPair); err != nil {
```
Comment (dmcgowan) [inline daemon/containerd/image_snapshot_unix.go:48]: Why does this only use the rootPair and not the full mapping?

Comment (rumpl) [reply daemon/containerd/image_snapshot_unix.go:48]: `RootPair` is the remapped root uid/gid

Comment (dmcgowan) [reply daemon/containerd/image_snapshot_unix.go:48]: Comparing to containerd's implementation which increments over the full userns range https://github.com/containerd/containerd/blob/v1.7.5/container_opts_unix.go#L104

Comment (rumpl) [reply daemon/containerd/image_snapshot_unix.go:48]: Ah right!

Comment (rumpl) [reply daemon/containerd/image_snapshot_unix.go:48]: Fixed thanks

File: daemon/containerd/image_snapshot_unix.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+			if err != nil {
+				return err
+			}
+
+			stat := info.Sys().(*syscall.Stat_t)
+			if stat == nil {
+				return fmt.Errorf("cannot get underlying data for %s", path)
+			}
+
+			return os.Lchown(path, int(stat.Uid)+idpair.UID, int(stat.Gid)+idpair.GID)
```
Comment (dmcgowan) [inline daemon/containerd/image_snapshot_unix.go:73]: Might be clearer to use `i.idMapping.ToContainer`

Comment (rumpl) [reply daemon/containerd/image_snapshot_unix.go]: Ok switched to using `ToHost`, `ToContainer` is the other way around and was giving me errors. With `ToHost` the uid/gid is the right one inside the container. Might be a good idea to add a test for this

Comment (rumpl) [reply daemon/containerd/image_snapshot_unix.go]: Added a test for this

File: daemon/containerd/image_snapshot_unix.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+		return err
+	}
+
+	mounts, err := snapshotter.Prepare(ctx, remappedID, parentSnapshot)
+	if err != nil {
+		return err
+	}
+
+	if err := i.remapRootFS(ctx, mounts); err != nil {
+		snapshotter.Remove(ctx, usernsID) //nolint: errcheck
```
Comment (vvoland) [inline daemon/containerd/image_snapshot_unix.go:49]: Could we log this error instead of ignoring it?

File: daemon/containerd/image_snapshot_windows.go [hunk 4]
```diff
@@ -0,0 +1,14 @@
+//go:build windows
```
Comment (vvoland) [inline daemon/containerd/image_snapshot_windows.go:1]: Not needed with `_windows` filename suffix

**Review by vvoland — APPROVED** [review]
Overall SGTM; left some nits though

File: daemon/containerd/image_snapshot_unix.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+		return err
+	}
+
+	mounts, err := snapshotter.Prepare(ctx, remappedID, parentSnapshot)
+	if err != nil {
+		return err
+	}
+
+	if err := i.remapRootFS(ctx, mounts); err != nil {
+		if rmErr := snapshotter.Remove(ctx, usernsID); err != nil {
```
Comment (vvoland) [inline daemon/containerd/image_snapshot_unix.go:50]:
```suggestion
		if rmErr := snapshotter.Remove(ctx, usernsID); rmErr != nil {
```

**Review by dmcgowan — APPROVED** [review]

File: daemon/snapshotter/mount.go [hunk 6]
```diff
@@ -93,7 +94,14 @@ func (m *refCountMounter) Mount(mounts []mount.Mount, containerID string) (targe
...
 
 	root := m.idMap.RootPair()
-	if err := idtools.MkdirAllAndChown(target, 0o700, root); err != nil {
+	dirID := idtools.Identity{
+		UID: idtools.CurrentIdentity().UID,
+		GID: root.GID,
+	}
+	if err := idtools.MkdirAllAndChown(path.Dir(target), 0o710, dirID); err != nil {
+		return "", err
+	}
```
Comment (thaJeztah) [inline daemon/snapshotter/mount.go:102]: Can you change this to use `filepath.Dir()` ?

I also thought for a long time that `path` is for `/` (and `filepath` only if you need cross-platform); https://github.com/golang/go/blob/2e7e1d2e8d96ddf2aff00feb645cfa94bc9e4786/src/path/path.go#L5-L6

> Package path implements utility routines for manipulating slash-separated paths.

But a second paragraph was added at some point! https://github.com/golang/go/commit/7465bfb1ea1a7ddfec9b267587ee9e6200514f3f

> To manipulate operating system paths, use the path/filepath package.

While at it, perhaps also inline that `dirID` var (I see the other implementations also have the above issues, but we can look at those separately)

```suggestion
	if err := idtools.MkdirAllAndChown(filepath.Dir(target), 0o710, idtools.Identity{
		UID: idtools.CurrentIdentity().UID,
		GID: root.GID,
	}); err != nil {
		return "", err
	}
```

File: daemon/containerd/image_snapshot_unix.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+				return err
+			}
+
+			stat := info.Sys().(*syscall.Stat_t)
+			if stat == nil {
+				return fmt.Errorf("cannot get underlying data for %s", path)
+			}
+
+			fid := idtools.Identity{UID: int(stat.Uid), GID: int(stat.Gid)}
+			ids, err := i.idMapping.ToHost(fid)
```
Comment (thaJeztah) [inline daemon/containerd/image_snapshot_unix.go:77]: Perhaps inline this one as well;

```suggestion
			ids, err := i.idMapping.ToHost(idtools.Identity{UID: int(stat.Uid), GID: int(stat.Gid)}
```

File: daemon/snapshotter/mount.go [hunk 8]
```diff
@@ -67,7 +67,6 @@ type refCountMounter struct {
 
 func (m *refCountMounter) Mount(mounts []mount.Mount, containerID string) (target string, retErr error) {
 	target = filepath.Join(m.home, mountsDir, m.snapshotter, containerID)
-
```
Comment (thaJeztah) [inline daemon/snapshotter/mount.go:69]: unrelated change? 🙈

**Review by thaJeztah — APPROVED** [review]
LGTM, but spotted some minor issues; let me know if you think it's worth fixing those before merging

File: daemon/snapshotter/mount.go (hunk 8 above)
Comment (rumpl) [reply daemon/snapshotter/mount.go:69]: Thought the code looked better that way

File: daemon/snapshotter/mount.go [hunk 9]
```diff
@@ -94,11 +93,10 @@ func (m *refCountMounter) Mount(mounts []mount.Mount, containerID string) (targe
 	}()
 
 	root := m.idMap.RootPair()
-	dirID := idtools.Identity{
+	if err := idtools.MkdirAllAndChown(filepath.Dir(target), 0o710, idtools.Identity{
 		UID: idtools.CurrentIdentity().UID,
 		GID: root.GID,
-	}
-	if err := idtools.MkdirAllAndChown(path.Dir(target), 0o710, dirID); err != nil {
+	}); err != nil {
```
[the remark this answers is not in the document]
Comment (rumpl) [reply daemon/snapshotter/mount.go:99]: :facepalm: forgot I had two commits in this PR

[the remark this answers is not in the document]
Comment (rumpl) [reply daemon/snapshotter/mount.go]: Fixed, thanks

File: daemon/containerd/image_snapshot.go [hunk 10]
```diff
@@ -63,17 +65,21 @@ func (i *ImageService) PrepareSnapshot(ctx context.Context, id string, parentIma
...
-	_, err = s.Prepare(ctx, id, parentSnapshot)
-	return err
+	snapshotter := i.client.SnapshotService(i.StorageDriver())
+
+	if i.idMapping.Empty() {
+		_, err = snapshotter.Prepare(ctx, id, parentSnapshot)
+		return err
+	}
+
+	return i.remapSnapshot(ctx, snapshotter, id, parentSnapshot, lease)
```
Comment (thaJeztah) [inline daemon/containerd/image_snapshot.go:82]: Ohman, changes look good, but I'm just looking at this part again, and this _could be_ a potential foot-gun if either a linter, or some "helpful" contribution fixes "unhandled error" here;

```suggestion

	if i.idMapping.Empty() {
		if _, err := snapshotter.Prepare(ctx, id, parentSnapshot); err != nil {
			return err
		}
	}

	return i.remapSnapshot(ctx, snapshotter, id, parentSnapshot, lease)
```


The above would break this code (which would _hopefully_ be caught in CI.

Do you think it's worth to be defensive here, and to reverse the order?

```suggestion

	if !i.idMapping.Empty() {
		return i.remapSnapshot(ctx, snapshotter, id, parentSnapshot, lease)
	}

	_, err = snapshotter.Prepare(ctx, id, parentSnapshot)
	return err
```

Comment (rumpl) [reply daemon/containerd/image_snapshot.go:82]: Sure I can reverse the order :+1:

**Review by thaJeztah — APPROVED** [review]
LGTM
